### PR TITLE
[extractor/dacast] Add extractors

### LINKED ISSUE #6163: Dacast.com
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2023.01.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

USA

### Example URLs

Single Video: https://iframe.dacast.com/vod/acae82153ef4d7a7344ae4eaa86af534/1c6143e3-5a06-371d-8695-19b96ea49090

### Provide a description that is worded well enough to be understood

Vzaar.com now is Dacast.com

Please support Dacast.com

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'test:dacast']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2023.01.06 [6becd25] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg N-109725-g2d202985b7-20230130 (setts), ffprobe N-109725-g2d202985b7-20230130
[debug] Optional libraries: Cryptodome-3.16.0, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1760 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2023.01.06, Current version: 2023.01.06
yt-dlp is up to date (2023.01.06)
[TestURL] Extracting URL: test:dacast
ERROR: [TestURL] None: No extractors matching 'dacast' found
  File "yt_dlp\extractor\common.py", line 680, in extract
  File "yt_dlp\extractor\testurl.py", line 24, in _real_extract
```

### PR DESCRIPTION
**Pull request by bashonly**
Adds VOD support for dacast.com, a French "streaming as a service" platform.

Livestreams and "expo" streams have different URL formats and are not supported with this PR.

Dacast purchased Vzaar in 2019, for which there is an extractor, and is in the process of sunsetting it. But some Vzaar links seem to still be alive and well, and are not redirecting to Dacast. Dead Vzaar links redirect to the Dacast home page.

Closes #6163

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/dacast.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    def _get_url_from_id(content_id):
+        url_type = 'playlist' if '-playlist-' in content_id else 'vod'
+        user_id, media_id = content_id.split(f'-{url_type}-')
+        return f'https://iframe.dacast.com/{url_type}/{user_id}/{media_id}'
+
+    @classmethod
+    def _extract_embed_urls(cls, url, webpage):
+        for content_id in re.findall(
+                r'<script[^>]+\bsrc=["\']https://player\.dacast\.com/js/player\.js\?contentId=([\w-]+-vod-[\w-]+)["\']', webpage):
+            yield DacastIE._get_url_from_id(content_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/dacast.py:39]: These could be in a base class:

```suggestion
class DacastBaseIE(InfoExtractor):
    @staticmethod
    def _get_url_from_id(content_id):
        url_type = 'playlist' if '-playlist-' in content_id else 'vod'
        user_id, media_id = content_id.split(f'-{url_type}-')
        return f'https://iframe.dacast.com/{url_type}/{user_id}/{media_id}'

    @classmethod
    def _extract_embed_urls(cls, url, webpage):
        for content_id in re.findall(
                r'<script[^>]+\bsrc=["\']https://player\.dacast\.com/js/player\.js\?contentId=([\w-]+-(?:vod|playlist)-[\w-]+)["\']', webpage):
            yield cls._get_url_from_id(content_id)
```

Not mandatory, just a suggestion

File: yt_dlp/extractor/dacast.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        if 'DRM_EXT' in hls_url:
+            self.report_drm(video_id)
+        elif '/uspaes/' in hls_url:
+            # From https://player.dacast.com/js/player.js
+            ts = int(time.time())
+            signature = hashlib.sha1(
+                f'{10413792000 - ts}{ts}YfaKtquEEpDeusCKbvYszIEZnWmBcSvw').digest().hex()
+            hls_aes['uri'] = f'https://keys.dacast.com/uspaes/{video_id}.key?s={signature}&ts={ts}'
+
+        formats = []
```
Comment (pukkandan) [inline yt_dlp/extractor/dacast.py:68]:
```suggestion
```

Unused. RetryManager will either succeed or raise since you haven't set fatal=False

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/dacast.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/dacast.py]: in order to move `_extract_embed_urls` to the base class, its regex pattern needs to be limited to 1 url type (`vod` vs `playlist`), or else embeds will be detected for both sub classes and one extractor will raise as unsuitable

Comment (pukkandan) [reply yt_dlp/extractor/dacast.py]: > in order to move `_extract_embed_urls` to the base class, its regex pattern needs to be limited to 1 url type (`vod` vs `playlist`), or else every embed will be detected by both sub classes and one extractor will raise as unsuitable

Nope. The results of _extract_embed_urls are tested again with _VALID_URL. But your solution works too.

Comment (bashonly) [reply yt_dlp/extractor/dacast.py]: I meant that it does this
```sh
[generic] Extracting URL: https://gist.githubusercontent.com/bashonly/7efb606f49f3c6e07ea0327de5a661d1/raw/05a16eac830245ea301fb0a585023bec71e6093c/dacast_playlist_embed.html
[generic] dacast_playlist_embed: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] dacast_playlist_embed: Extracting information
[debug] Looking for embeds
[debug] Identified a Dacast embed
[debug] Identified a DacastPlaylist embed
[download] Downloading playlist: hello
[generic] Playlist hello: Downloading 2 items of 2
[download] Downloading item 1 of 2
ERROR: No suitable extractor (Dacast) found for URL https://iframe.dacast.com/playlist/943bb1ab3c03695ba85330d92d6d226e/b632eb053cac17a9c9a02bcfc827f2d8
  File "/home/user/bin/yt_dlp", line 9, in <module>
    sys.exit(main())
  File "/home/user/git/yt-dlp/yt_dlp/__init__.py", line 969, in main
    _exit(*variadic(_real_main(argv)))
  File "/home/user/git/yt-dlp/yt_dlp/__init__.py", line 959, in _real_main
    return ydl.download(all_urls)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 3369, in download
    self.__download_wrapper(self.extract_info)(
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 3344, in wrapper
    res = func(*args, **kwargs)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1507, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1518, in wrapper
    return func(self, *args, **kwargs)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1615, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1744, in process_ie_result
    return self.__process_playlist(ie_result, download)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1888, in __process_playlist
    entry_result = self.__process_iterable_entry(entry, download, collections.ChainMap({
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1518, in wrapper
    return func(self, *args, **kwargs)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1920, in __process_iterable_entry
    return self.process_ie_result(
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1694, in process_ie_result
    return self.extract_info(
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1510, in extract_info
    self.report_error(f'No suitable extractor{format_field(ie_key, None, " (%s)")} found for URL {url}',
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 1015, in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
  File "/home/user/git/yt-dlp/yt_dlp/YoutubeDL.py", line 944, in trouble
    tb_data = traceback.format_list(traceback.extract_stack())

[download] Downloading item 2 of 2
[DacastPlaylist] Extracting URL: https://iframe.dacast.com/playlist/943bb1ab3c03695ba85330d92d6d226e/b632eb053cac17a9c9a02bcfc827f2d8
[DacastPlaylist] b632eb053cac17a9c9a02bcfc827f2d8: Downloading JSON metadata
[download] Downloading playlist: Archive Sermons
[DacastPlaylist] Playlist Archive Sermons: Downloading 29 items of 29
[download] Downloading item 1 of 29
[Dacast] Extracting URL: https://iframe.dacast.com/vod/943bb1ab3c03695ba85330d92d6d226e/dfb24953-bd3d-42a5-8d8c-a55f91899595
[Dacast] dfb24953-bd3d-42a5-8d8c-a55f91899595: Downloading JSON metadata
[Dacast] dfb24953-bd3d-42a5-8d8c-a55f91899595: Downloading access JSON
^C
ERROR: Interrupted by user
```

it does download the embed, but I'd rather it didn't error first

**Comment by pukkandan** [comment]
There are no _WEBPAGE_TESTS ?

**Review by pukkandan — APPROVED** [review]

**Comment by bashonly** [comment]
> There are no _WEBPAGE_TESTS ?

My bad. It's surprisingly hard to find webpages with Dacast embeds. I've added the example embed from the Dacast support page as a test, and then a couple of pages I mocked up using the embed code that Dacast provides after uploading a video

**Comment by bashonly** [comment]
Was the change to `_extract_embed_urls` not good? Or should I not do the homemade tests?

**Pull request merged**